src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -174,6 +169,36 @@ entails SolverOptions{..} constraint context hints =
     forClassName ctx cn@(Qualified (Just mn) _) tys = concatMap (findDicts ctx cn) (ordNub (Nothing : Just mn : map Just (mapMaybe ctorModules tys)))
     forClassName _ _ _ = internalError "forClassName: expected qualified class name"
 
+    -- Right biased union of two row types
+    unionRows :: Type -> Type -> Type -> Maybe (Type, Maybe [Constraint])
+    unionRows l r o =
+        let (ls, lEnd) = rowToList l
+            (rs, rEnd) = rowToList r
+            (os, oEnd) = rowToList o
+            lm = M.fromList ls
+            rm = M.fromList rs
+            lrUnion = M.toList $ M.union rm lm
+            unioned = rowFromList (lrUnion, oEnd)
+            constraintDiff a b m = Just $ [ Constraint C.RowUnion [a, b, rowFromList (M.toList $ M.difference (M.fromList os) m, oEnd)] Nothing ]
+        in case (isClosedRow lEnd, isClosedRow rEnd) of
```

Comment:
I would have expected that we could just union the known labels and then add a `RowUnion` constraint on the tails, regardless of what the tails are. We could provide an instance for `RowUnion () () ()` with the class.

Also bear in mind  that the tail could be a skolem constant, which is not being handled here. If you do it this way, you won't need any special cases.